### ISSUE TITLE
/clr:implicitKeepAlive- is not documented

### ISSUE THREAD

**Issue by mwinterb** [MicrosoftDocs/cpp-docs#3177]
The command line help for cl.exe in VS2019 v16.10 mentions implicitKeepAlive- as an option, but it is not yet documented:
```
/clr[:option] compile for common language runtime, where option is:
    pure : produce IL-only output file (no native executable code)
    safe : produce IL-only verifiable output file
    netcore : produce assemblies targeting .NET Core runtime
    noAssembly : do not produce an assembly
    nostdlib : ignore the system .NET framework directory when searching for assemblies
    nostdimport : do not import any required assemblies implicitly
    initialAppDomain : enable initial AppDomain behavior of Visual C++ 2002
    implicitKeepAlive- : turn off implicit emission of System::GC::KeepAlive(this)
```
The documentation provided in the command line help is adequate for what the option does and could probably just be cloned here, but what's missing is "in what cases does the compiler automatically make calls to `System::GC::KeepAlive(this)`"?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: a1d1e8ce-e0d5-321a-1cd3-8d96fecde8ad
* Version Independent ID: 04343afb-ac3b-a7cb-6322-f5608442f554
* Content: [/clr (Common Language Runtime compilation)](https://docs.microsoft.com/en-us/cpp/build/reference/clr-common-language-runtime-compilation?view=msvc-160)
* Content Source: [docs/build/reference/clr-common-language-runtime-compilation.md](https://github.com/Microsoft/cpp-docs/blob/master/docs/build/reference/clr-common-language-runtime-compilation.md)
* Product: **visual-cpp**
* Technology: **cpp-tools**
* GitHub Login: @corob-msft
* Microsoft Alias: **corob**

**Comment by corob-msft** [comment]
@mwinterb 
Good question. The compiler behavior has changed in 16.10 to be more defensive against JIT GC operations, which became more aggressively optimized in .NET Framework 4.8 (and possibly in .NET Core as well), potentially causing race conditions that were less common previously. For now, I have this guidance on when we emit `System::GC::KeepAlive` calls automatically, which is when this set of conditions is satisfied:
- We're returning from a member function of a managed class,
- The managed class has a member of type 'reference or pointer to an unmanaged class-type', including members in its direct or indirect base classes, or embedded in members of class-types, occurring anywhere in the class hierarchy,
- A call to an unmanaged function is found in the current (managed member) function.

Basically, it's emitted whenever you'd be likely to do the same thing by hand yourself, if you were aware of the issue. The one difference is that the call is emitted during IL generation a bit later in the function return process than it would be if made in the source code. Once again, this is to prevent another race condition.

There's also a new `#pragma` directive to selectively enable or disable this behavior, if the global compiler switch is too broad:

> **`#pragma implicit_keepalive (`** [ **`push ,`** ] **`on )`**\
> **`#pragma implicit_keepalive (`** [ **`push ,`** ]  **`off )`**\
> **`#pragma implicit_keepalive ( pop )`**

Note to self: tracking bug 1229778, PR 316163

**Comment by Alois-xx** [comment]
@corob-msft: Can you elaborate the "other" race condition upon method return? 
Will this issue manifest mainly as early GCed parent objects whose members dtors are called too early leading to NullReferenceExceptions or are also Managed Heap Corruptions expected?

**Comment by corob-msft** [comment]
@Alois-xx 
My description of return handling may have been too broad. The compiler's placement of the `KeepAlive` call isn't magic relative to what's *possible* in source code. However, it may insert it later relative to what's found in the *actual* source code passed to the compiler. This is what the developer wrote about placement of the `KeepAlive` call relative to the return statement:
> If we emit a `System::GC::KeepAlive(this)` call immediately preceding the return it still leaves
> opportunity for the JIT compiler to deallocate `this` before the call to `copyData()` is complete.
> 
> E.g.,
> 
```
auto Bridge::CallLong()
{
    return this->mPtr->copyData();
}
```
> Fixing this edge case requires rewriting the return expression and inserting the `KeepAlive` call
> at the appropriate place but doing this at the tree level proved impossible due to the penchant
> for tree pattern matching in the compiler. This is instead done at IL Lower level.
So in code that looks like this example, a naive approach that just inserts a `KeepAlive` call before the `return` isn't sufficient to avoid the race. You may need to refactor your code slightly to prevent a possible GC during processing of your return statement. That's essentially what the compiler does here. 

As for how the premature GC issue manifests, I'd expect null reference exceptions. I don't know enough to say you *can't* get managed heap corruption, but that's not the usual result of running finalizers.

**Comment by Alois-xx** [comment]
@corob-msft : Thanks for the clarification.  

The race condition comes from finalizers but if Bridge does not contain a finalizer GC.KeepAlive is also emitted as you noted by the rules above. This means that if a managed class has a member variable which holds some finalizable object then you also have a race condition which is prevented by this fix. 
It all boils down to executing finalizers for objects which unmanaged method call still might be needing. 

This would then show up at runtime as invalid handles, nulled out memory, already released memory, closed sockets, files, ...  and such things.

**Comment by TylerMSFT** [comment]
Thank you for your patience as we work through our backlog. We have received more issues than we can manage at this time. We are closing issues based on their severity, how broadly applicable they are, and whether there is enough information provided to act on the issue. If you believe that this issue should not have been closed, please add a comment explaining why by 3/1/2024. We appreciate your contribution and encourage you to propose changes using public PRs, or submit feedback through our updated control at the top of the page, as we phase out Git Issues for this repo. For more information on our feedback systems, see https://aka.ms/ContentUserFeedback.
